### PR TITLE
Project restructure to make it easier to implement more backends

### PR DESCRIPTION
**Pull request by sritchie**
This PR refactors:

- `caliban.config` into `caliban.config` and `caliban.config.experiment`
- `caliban.util` becomes `caliban.util` and `caliban.util.{argparse, auth, fs, tqdm}` 
- `caliban.cloud.*` becomes `caliban.platform.cloud.*`
- `caliban.docker` becomes `caliban.docker.{build, push}` and `caliban.platform.{run, shell, notebook}`

The tests are also split accordingly.

I snuck in ONE bugfix that appeared during testing. I moved our method of generating credentials from `gcloud auth login` over to `caliban.util.auth` so that I could reuse it in the caliban history API. We had one spot where we were calling `discovery.build` and NOT passing any credentials, which threw an error when the user didn't have `GOOGLE_APPLICATION_CREDENTIALS` specified.

I also removed unused imports and sorted imports wherever I could.

## Why?

The project needed it; specifically this is going to make it easier to increase test coverage, and to get the `.calibanconfig.json` code working with a proper schema, so that we can start to add options there.

I really want to make the platforms pluggable, and sorting them out into separate modules is the first step.

### CODE REVIEW

**Comment by sritchie** [comment]
@ajslone , I've tested `caliban {shell, run, cloud}`, and `caliban status` just with the basic command, no options. Would you mind helping me test the GKE code, and fixing any busted imports that I've forgotten?

Thank you!

File: caliban/history/util.py [hunk 1]
```diff
@@ -244,8 +244,12 @@ def _get_caip_job_name(j: Job) -> str:
 
 
 # ----------------------------------------------------------------------------
-def _get_caip_job_api() -> Any:
-  return discovery.build('ml', 'v1', cache_discovery=False).projects().jobs()
+def _get_caip_job_api(credentials_path: Optional[str] = None) -> Any:
```
Comment (ajslone) [inline caliban/history/util.py:247]: Nice addition here.

File: caliban/main.py [hunk 2]
```diff
@@ -18,20 +18,22 @@
...
 import caliban.cli as cli
-import caliban.cloud.core as cloud
 import caliban.config as c
-import caliban.docker as docker
-import caliban.gke as gke
-import caliban.gke.cli
-import caliban.util as u
+import caliban.docker.build as b
 import caliban.history.cli
+import caliban.platform.cloud.core as cloud
```
Comment (ajslone) [inline caliban/main.py:30]: Definitely like the platform.xxx structure better here.

File: caliban/platform/cloud/core.py [hunk 3]
```diff
@@ -279,17 +261,7 @@ def ml_api(credentials_path: Optional[str] = None):
...
-  if credentials_path is not None:
-    credentials = service_account.Credentials.from_service_account_file(
-        credentials_path)
-  else:
-    # attempt to fetch credentials acquired via `gcloud auth login`. If this
-    # fails, the following API object will attempt to use application default
-    # credentials.
-    credentials = gcloud_auth_credentials()
-
+  credentials = ua.gcloud_credentials(credentials_path)
```
Comment (ajslone) [inline caliban/platform/cloud/core.py:264]: cleaner, nice

**Pull request merged**